optuna/integration/allennlp.py

```diff
@@ -9,17 +10,156 @@
+    the one's configuration could be loaded in the another's configuration.
+    To avoid this hazard, we add ID of a parent process to each key of
+    environment variables.
+
+    """
+    ppid = os.getppid()
+    return "{}_{}".format(ppid, key)
+
+
+def _create_pruner() -> Optional[optuna.pruners.BasePruner]:
+    """Restore a pruner which is defined in `create_study`.
+
+    `AllenNLPPruningCallback` is launched as a sub-process of
+    a main script that defines search spaces.
+    An instance cannot be passed directly from the parent process
+    to its sub-process. For this reason, we set information about
+    pruner as environment variables and load them and
+    re-create the same pruner in `AllenNLPPruningCallback`.
+
+    """
+    pruner_class = os.getenv(_add_ppid("OPTUNA_ALLENNLP_PRUNER_CLASS"))
+    if pruner_class is None:
+        return None
+
+    pruner_params = _get_environment_variables_for_pruner()
+    pruner = getattr(optuna.pruners, pruner_class, None)
+
+    if pruner is None:
+        return None
+
+    return pruner(**pruner_params)
+
+
+def _infer_and_cast(value: Optional[str]) -> Optional[Union[str, int, float, bool]]:
+    """Infer and cast a string to desired types.
+
+    We are only able to set strings as environment variables.
+    However, parameters of a pruner could be integer, float,
+    boolean, or else. We infer and cast environment variables
+    to desired types.
+
+    """
+    if value is None:
+        return None
+
+    try:
+        return int(value)
+    except ValueError:
+        try:
+            return float(value)
+        except ValueError:
+            if value == "True":
+                return True
+            elif value == "False":
+                return False
+
+    return value
+
+
+def _get_environment_variables_for_trial() -> Dict[str, Optional[str]]:
+    return {
+        "study_name": os.getenv(_add_ppid("OPTUNA_ALLENNLP_STUDY_NAME")),
+        "trial_id": os.getenv(_add_ppid("OPTUNA_ALLENNLP_TRIAL_ID")),
+        "storage": os.getenv(_add_ppid("OPTUNA_ALLENNLP_STORAGE_NAME")),
+        "monitor": os.getenv(_add_ppid("OPTUNA_ALLENNLP_MONITOR")),
+    }
+
+
+def _get_environment_variables_for_pruner() -> Dict[str, Optional[Union[str, int, float, bool]]]:
+    keys = os.getenv(_add_ppid("OPTUNA_ALLENNLP_PRUNER_KEYS"))
+    if keys is None:
+        return {}
+
+    kwargs = {}
+    for key in keys.split(","):
+        parameter_key = "OPTUNA_ALLENNLP_{}".format(key)
+        value = os.getenv(_add_ppid(parameter_key))
+        kwargs[key] = _infer_and_cast(value)
+
+    return kwargs
+
+
+def _fetch_pruner_config(trial: optuna.Trial) -> Dict[str, Any]:
+    pruner = trial.study.pruner
+    kwargs = {}  # type: Dict[str, Any]
+
+    if isinstance(pruner, optuna.pruners.HyperbandPruner):
+        kwargs["min_resource"] = pruner._min_resource
+        kwargs["max_resource"] = pruner._max_resource
+        kwargs["reduction_factor"] = pruner._reduction_factor
```

Comment:
`HyperbandPruner` has some other states including `_pruners` and `_total_trial_allocation_budget` (https://github.com/optuna/optuna/blob/master/optuna/pruners/_hyperband.py#L143-L146). I guess it initializes the successive halving pruner every trial. Do you think if it affects optimization performance?

Reply:
For example, the estimation of `max_resource` cannot be estimated correctly if the value of `max_resource` is `auto` and the resource (epoch) of each trial is different.

But, maybe this is an issue of `HyperbandPruner` and should be resolved by the pruner.